src/NexusMods.DataModel/GameRegistry/Registry.cs

```diff
@@ -19,110 +23,141 @@ namespace NexusMods.DataModel;
     {
-        return (installation.Game.Domain, installation.Version, installation.Store);
+        var found = db
+            .FindIndexed(result.Path.ToString(), GameMetadata.Path)
+            .Select(db.Get<GameMetadata.Model>)
+            .FirstOrDefault(m => m.Domain == locatableGame.Domain && m.Store == result.Store);
+        if (found is null)
+        {
+            id = null;
+            return false;
+        }
+        id = found.Id;
+        return true;
     }
 
-    /// <inheritdoc />
-    public IEnumerable<GameInstallation> AllInstalledGames => _byId.Values;
+    /// <summary>
+    /// Gets the locator id from the database, if none exists it will be created and returned.
+    /// </summary>
+    private async ValueTask<EntityId> GetLocatorId(ILocatableGame locatableGame, GameLocatorResult result)
+    {
+        if (TryGetLocatorResultId(_conn.Db, locatableGame, result, out var id))
+            return id.Value;
+        
+        using var tx = _conn.BeginTransaction();
+        tx.Add(locatableGame, result, static (tx, db, game, result) =>
+        {
+            // Check for a race condition, someone may have added it before us.
+            if (TryGetLocatorResultId(db, game, result, out var _))
+                return;
 
-    /// <inheritdoc />
-    public GameInstallation Get(EntityId id) => _byId[id];
+            // Doesn't exist, so create it.
+            _ = new GameMetadata.Model(tx)
+            {
+                Store = result.Store.Value,
+                Domain = game.Domain.Value,
+                Path = result.Path.ToString(),
+            };
+        });
+        
+        var txResult = await tx.Commit();
+        
+        if (!TryGetLocatorResultId(txResult.Db, locatableGame, result, out id))
+            throw new InvalidOperationException("Failed to get locator result id after inserting it, this should never happen");
+        
+        return id.Value;
+    }
 
-    /// <inheritdoc />
-    public EntityId GetId(GameInstallation installation) => _byInstall[GetKey(installation)];
 
+    /// <summary>
+    /// Looks through all locators to find all installations.
+    /// </summary>
+    private async IAsyncEnumerable<GameInstallation> FindInstallations()
+    {
+        foreach (var game in _games)
+        {
+            foreach (var locator in _locators)
+            {
+                foreach (var found in locator.Find(game))
+                {
+                    yield return await Register(game, found, locator);
+                }
+            }
+        }
+    }
+    
     /// <inheritdoc />
     public async Task StartAsync(CancellationToken cancellationToken)
     {
-        await ((IHostedService)_conn).StartAsync(cancellationToken);
-        await Startup(_games);
+        lock (this)
+        {
+            _startupTask ??= Startup(cancellationToken);
+        }
+        await _startupTask;
```

Comment:
Why are we locking here?

Reply:
There's a race condition here: 

C# example:

```csharp
using System;
public class C {
    int? i = null;
    public int M(int v) {
        i ??= v;
        return i.Value;
    }
}
```

compiles to

```csharp
using System;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Security.Permissions;

[assembly: CompilationRelaxations(8)]
[assembly: RuntimeCompatibility(WrapNonExceptionThrows = true)]
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints)]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, SkipVerification = true)]
[assembly: AssemblyVersion("0.0.0.0")]
[module: UnverifiableCode]
[module: RefSafetyRules(11)]

public class C
{
    private Nullable<int> i;

    public int M(int v)
    {
        int valueOrDefault = i.GetValueOrDefault();
        if (!i.HasValue)
        {
            valueOrDefault = v;
            i = valueOrDefault;
        }
        return i.Value;
    }
}
```

So there's a small chance that wo threads may both try and start/set the value at the same time, the lock removes this chance.